backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/attribute/impl/HUUniqueAttributesRepository.java

```diff
@@ -0,0 +1,136 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.handlingunits.attribute.impl;
+
+import de.metas.handlingunits.HuId;
+import de.metas.handlingunits.HuPackingInstructionsAttributeId;
+import de.metas.handlingunits.model.I_M_HU_Attribute;
+import de.metas.handlingunits.model.I_M_HU_UniqueAttribute;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.ad.dao.IQueryBL;
+import org.adempiere.ad.dao.IQueryBuilder;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+import static org.adempiere.model.InterfaceWrapperHelper.newInstance;
+import static org.adempiere.model.InterfaceWrapperHelper.save;
+
+@Repository
+public class HUUniqueAttributesRepository
+{
+	final IQueryBL queryBL = Services.get(IQueryBL.class);
+
+	public void createOrUpdateHUUniqueAttribute(@NonNull final HUUniqueAttributeCreateRequest huUniqueAttributeCreateRequest)
+	{
+		final I_M_HU_UniqueAttribute existingHUUniqueAttribute = retrieveHUUniqueAttributeForHU(huUniqueAttributeCreateRequest.getHuId());
+
+		if (existingHUUniqueAttribute != null)
+		{
+			// hu was unique attribute was already recorded. Update the value if needed
+			if (!huUniqueAttributeCreateRequest.getAttributeValue().equals(existingHUUniqueAttribute.getValue()))
+			{
+				existingHUUniqueAttribute.setValue(huUniqueAttributeCreateRequest.getAttributeValue());
+				save(existingHUUniqueAttribute);
+			}
+
+			return;
+		}
+
+		final I_M_HU_UniqueAttribute huUniqueAttributeRecord = newInstance(I_M_HU_UniqueAttribute.class);
+
+		huUniqueAttributeRecord.setM_HU_PI_Attribute_ID(huUniqueAttributeCreateRequest.getHuPIAttributeId().getRepoId());
+		huUniqueAttributeRecord.setM_HU_Attribute_ID(huUniqueAttributeCreateRequest.getHuAttributeId());
+		huUniqueAttributeRecord.setM_HU_ID(huUniqueAttributeCreateRequest.getHuId().getRepoId());
+		huUniqueAttributeRecord.setM_Product_ID(huUniqueAttributeCreateRequest.getProductId().getRepoId());
+		huUniqueAttributeRecord.setM_Attribute_ID(huUniqueAttributeCreateRequest.getAttributeId().getRepoId());
+		huUniqueAttributeRecord.setValue(huUniqueAttributeCreateRequest.getAttributeValue());
+
+		save(huUniqueAttributeRecord);
+	}
+
+	private I_M_HU_UniqueAttribute retrieveHUUniqueAttributeForHU(@NonNull final HuId huId)
+	{
+		return queryBL.createQueryBuilder(I_M_HU_UniqueAttribute.class)
+				.addEqualsFilter(I_M_HU_UniqueAttribute.COLUMNNAME_M_HU_ID, huId)
+				.create()
+				.firstOnly(I_M_HU_UniqueAttribute.class);
+	}
+
+	public I_M_HU_UniqueAttribute retrieveHUUniqueAttributeForProductAndAttributeValue(@NonNull final HUUniqueAttributeParameters huUniqueAttributeParameters)
```

Comment:
i don't think we can or should (or have to) assume that an HU has just one unique attribute ❗ 
Afaiu it makes sense to select using M_HU_ID and M_Attribute_ID

Reply:
I will also add the product in the query. I don't know if the storages could contain different products for the same HU but codewise they could.